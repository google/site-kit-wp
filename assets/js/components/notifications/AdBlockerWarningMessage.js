/**
 * AdBlockerWarningMessage component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { sprintf, __ } from '@wordpress/i18n';
import { createInterpolateElement } from '@wordpress/element';

/**
 * Internal dependencies
 */
import Link from '../Link';
import WarningNotice from '../WarningNotice';
import ExternalIcon from '../../../svg/icons/external-rounded.svg';

export default function AdBlockerWarningMessage( {
	className = '',
	getHelpLink = '',
	warningMessage = null,
} ) {
	if ( ! warningMessage ) {
		return null;
	}

	return (
		<WarningNotice className={ className }>
			{ createInterpolateElement(
				sprintf(
					/* translators: 1: The warning message. 2: "Get help" text. */
					__( '%1$s. <Link>%2$s</Link>', 'google-site-kit' ),
					warningMessage,
					__( 'Get help', 'google-site-kit' )
				),
				{
					Link: (
						<Link
							href={ getHelpLink }
							external
							hideExternalIndicator
							trailingIcon={
								<ExternalIcon width={ 15 } height={ 15 } />
							}
						/>
					),
				}
			) }
		</WarningNotice>
	);
}

AdBlockerWarningMessage.propTypes = {
	className: PropTypes.string,
	getHelpLink: PropTypes.string,
	warningMessage: PropTypes.string,
};
