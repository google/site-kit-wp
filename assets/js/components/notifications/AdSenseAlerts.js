/**
 * AdSenseAlerts component.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { Fragment } from '@wordpress/element';
import { __, _x } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { CORE_MODULES } from '../../googlesitekit/modules/datastore/constants';
import { MODULES_ADSENSE } from '../../modules/adsense/datastore/constants';
import NotificationAlertSVG from '../../../svg/graphics/notification-alert.svg';
import BannerNotification from '../notifications/BannerNotification';

function AdSenseAlerts() {
	const adSenseModuleConnected = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleConnected( 'adsense' )
	);
	const accountID = useSelect( ( select ) =>
		select( MODULES_ADSENSE ).getAccountID()
	);
	const notifications = useSelect( ( select ) =>
		select( MODULES_ADSENSE ).getNotifications()
	);

	if (
		! adSenseModuleConnected ||
		! accountID ||
		notifications === undefined
	) {
		return null;
	}

	return (
		<Fragment>
			{ notifications.map(
				( {
					id,
					title,
					description,
					format,
					ctaURL,
					ctaLabel,
					ctaTarget,
					severity,
					isDismissable,
				} ) => (
					<BannerNotification
						className="googlesitekit-adsense-alert"
						key={ id }
						id={ id }
						title={ title || '' }
						description={ description }
						WinImageSVG={ NotificationAlertSVG }
						format={ format || 'small' }
						ctaLink={ ctaURL }
						ctaLabel={ ctaLabel }
						ctaTarget={ ctaTarget }
						type={ severity }
						dismiss={ __( 'OK, Got it!', 'google-site-kit' ) }
						isDismissible={ isDismissable || true }
						logo
						module="adsense"
						moduleName={ _x(
							'AdSense',
							'Service name',
							'google-site-kit'
						) }
						dismissExpires={ 0 }
						showOnce={ false }
					/>
				)
			) }
		</Fragment>
	);
}

export default AdSenseAlerts;
