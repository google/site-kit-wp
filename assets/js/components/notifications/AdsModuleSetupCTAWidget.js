/**
 * AdsModuleSetupCTAWidget component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { compose } from '@wordpress/compose';
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useDispatch, useSelect } from 'googlesitekit-data';
import { Button } from 'googlesitekit-components';
import {
	BREAKPOINT_SMALL,
	BREAKPOINT_TABLET,
	useBreakpoint,
} from '../../hooks/useBreakpoint';
import { ADS_MODULE_SETUP_BANNER_PROMPT_DISMISSED_KEY } from '../../modules/ads/datastore/constants';
import { CORE_USER } from '../../googlesitekit/datastore/user/constants';
import { WEEK_IN_SECONDS } from '../../util';
import { Cell, Grid, Row } from '../../material-components';
import whenInactive from '../../util/when-inactive';
import { withWidgetComponentProps } from '../../googlesitekit/widgets/util';
import useActivateModuleCallback from '../../hooks/useActivateModuleCallback';
import AdsSetupSVG from '../../../svg/graphics/ads-setup.svg';
import AdsSetupTabletSVG from '../../../svg/graphics/ads-setup-tablet.svg';
import AdsSetupMobileSVG from '../../../svg/graphics/ads-setup-mobile.svg';

function AdsModuleSetupCTAWidget( { WidgetNull, Widget } ) {
	const breakpoint = useBreakpoint();
	const isMobileBreakpoint = breakpoint === BREAKPOINT_SMALL;
	const isTabletBreakpoint = breakpoint === BREAKPOINT_TABLET;

	const onSetupCallback = useActivateModuleCallback( 'ads' );

	const isDismissed = useSelect( ( select ) =>
		select( CORE_USER ).isPromptDismissed(
			ADS_MODULE_SETUP_BANNER_PROMPT_DISMISSED_KEY
		)
	);
	const dismissCount = useSelect( ( select ) =>
		select( CORE_USER ).getPromptDismissCount(
			ADS_MODULE_SETUP_BANNER_PROMPT_DISMISSED_KEY
		)
	);

	const { dismissPrompt } = useDispatch( CORE_USER );

	const onDismiss = useCallback( async () => {
		if ( dismissCount < 1 ) {
			const twoWeeksInSeconds = WEEK_IN_SECONDS * 2;

			await dismissPrompt( ADS_MODULE_SETUP_BANNER_PROMPT_DISMISSED_KEY, {
				expiresInSeconds: twoWeeksInSeconds,
			} );
		} else {
			// For the second dismissal, dismiss permanently.
			await dismissPrompt( ADS_MODULE_SETUP_BANNER_PROMPT_DISMISSED_KEY );
		}
	}, [ dismissCount, dismissPrompt ] );

	if ( dismissCount > 1 || isDismissed || isDismissed === undefined ) {
		return <WidgetNull />;
	}

	return (
		<div className="googlesitekit-widget-context">
			<Grid className="googlesitekit-widget-area">
				<Row>
					<Cell size={ 12 }>
						<Widget
							noPadding
							className="googlesitekit-setup-cta-banner googlesitekit-ads-setup-cta-widget"
						>
							<Grid collapsed>
								<Row>
									<Cell
										smSize={ 12 }
										mdSize={ 8 }
										lgSize={ 6 }
										className="googlesitekit-setup-cta-banner__primary-cell"
									>
										<h3 className="googlesitekit-setup-cta-banner__title">
											{ __(
												'Get better quality leads and enhance conversions with Ads',
												'google-site-kit'
											) }
										</h3>
										<div className="googlesitekit-setup-cta-banner__description">
											<p>
												{ __(
													'Help drive sales, leads, or site traffic by getting your business in front of people who are actively searching Google for products or services you offer.',
													'google-site-kit'
												) }
											</p>
										</div>

										<div className="googlesitekit-setup-cta-banner__actions-wrapper">
											<Button
												className="googlesitekit-key-metrics-cta-button"
												onClick={ onSetupCallback }
											>
												{ __(
													'Set up Ads',
													'google-site-kit'
												) }
											</Button>
											<Button
												tertiary
												onClick={ onDismiss }
											>
												{ dismissCount < 1
													? __(
															'Maybe later',
															'google-site-kit'
													  )
													: __(
															'Donâ€™t show again',
															'google-site-kit'
													  ) }
											</Button>
										</div>
									</Cell>
									{ ! isMobileBreakpoint &&
										! isTabletBreakpoint && (
											<Cell
												alignBottom
												className="googlesitekit-setup-cta-banner__svg-wrapper"
												mdSize={ 8 }
												lgSize={ 6 }
											>
												<AdsSetupSVG />
											</Cell>
										) }
									{ isTabletBreakpoint && (
										<Cell
											className="googlesitekit-setup-cta-banner__svg-wrapper"
											mdSize={ 8 }
										>
											<AdsSetupTabletSVG />
										</Cell>
									) }
									{ isMobileBreakpoint && (
										<Cell
											alignBottom
											className="googlesitekit-setup-cta-banner__svg-wrapper"
											smSize={ 12 }
										>
											<AdsSetupMobileSVG />
										</Cell>
									) }
								</Row>
							</Grid>
						</Widget>
					</Cell>
				</Row>
			</Grid>
		</div>
	);
}

AdsModuleSetupCTAWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	WidgetNull: PropTypes.elementType.isRequired,
};

export default compose(
	whenInactive( { moduleName: 'ads' } ),
	withWidgetComponentProps( 'ads-setup-cta' )
)( AdsModuleSetupCTAWidget );
