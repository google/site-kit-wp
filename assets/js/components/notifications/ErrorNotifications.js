/**
 * ErrorNotifications component.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { Fragment } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import AuthError from './AuthError';
import InternalServerError from './InternalServerError';
import {
	CORE_USER,
	FORM_TEMPORARY_PERSIST_PERMISSION_ERROR,
} from '../../googlesitekit/datastore/user/constants';
import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
import { CORE_FORMS } from '../../googlesitekit/datastore/forms/constants';
import BannerNotification from './BannerNotification';
import Notifications from './Notifications';
import { NOTIFICATION_AREAS } from '../../googlesitekit/notifications/datastore/constants';

export default function ErrorNotifications() {
	const isAuthenticated = useSelect( ( select ) =>
		select( CORE_USER ).isAuthenticated()
	);

	// These will be `null` if no errors exist.
	const setupErrorCode = useSelect( ( select ) =>
		select( CORE_SITE ).getSetupErrorCode()
	);
	const setupErrorMessage = useSelect( ( select ) =>
		select( CORE_SITE ).getSetupErrorMessage()
	);
	const temporaryPersistedPermissionsError = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue(
			FORM_TEMPORARY_PERSIST_PERMISSION_ERROR,
			'permissionsError'
		)
	);
	const setupErrorRedoURL = useSelect( ( select ) => {
		if ( temporaryPersistedPermissionsError?.data ) {
			return select( CORE_USER ).getConnectURL( {
				additionalScopes:
					temporaryPersistedPermissionsError?.data?.scopes,
				redirectURL:
					temporaryPersistedPermissionsError?.data?.redirectURL ||
					global.location.href,
			} );
		} else if (
			setupErrorCode === 'access_denied' &&
			! temporaryPersistedPermissionsError?.data &&
			isAuthenticated
		) {
			return null;
		}

		return select( CORE_SITE ).getSetupErrorRedoURL();
	} );
	const errorTroubleshootingLinkURL = useSelect( ( select ) =>
		select( CORE_SITE ).getErrorTroubleshootingLinkURL( {
			code: setupErrorCode,
		} )
	);

	let title = __( 'Error connecting Site Kit', 'google-site-kit' );
	let ctaLabel = __( 'Redo the plugin setup', 'google-site-kit' );

	if ( setupErrorCode === 'access_denied' ) {
		title = __( 'Permissions Error', 'google-site-kit' );

		if ( temporaryPersistedPermissionsError?.data ) {
			ctaLabel = __( 'Grant permission', 'google-site-kit' );
		} else if (
			! temporaryPersistedPermissionsError?.data &&
			isAuthenticated
		) {
			ctaLabel = null;
		}
	}

	if (
		temporaryPersistedPermissionsError?.data?.skipDefaultErrorNotifications
	) {
		return null;
	}

	return (
		<Fragment>
			<InternalServerError />
			<AuthError />
			{ setupErrorMessage && (
				<BannerNotification
					id="setup_error"
					type="win-error"
					title={ title }
					description={ setupErrorMessage }
					isDismissible={ false }
					ctaLink={ setupErrorRedoURL }
					ctaLabel={ ctaLabel }
					learnMoreLabel={ __( 'Get help', 'google-site-kit' ) }
					learnMoreURL={ errorTroubleshootingLinkURL }
				/>
			) }
			<Notifications areaSlug={ NOTIFICATION_AREAS.ERRORS } />
		</Fragment>
	);
}
