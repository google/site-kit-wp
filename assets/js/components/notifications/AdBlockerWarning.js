/**
 * AdSense AdBlockerWarning component.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
import { CORE_MODULES } from '../../googlesitekit/modules/datastore/constants';
import AdBlockerWarningMessage from './AdBlockerWarningMessage';

export default function AdBlockerWarning( { moduleSlug, className } ) {
	const storeName = useSelect( ( select ) =>
		select( CORE_MODULES ).getModuleStoreName( moduleSlug )
	);
	const adBlockerWarningMessage = useSelect( ( select ) =>
		select( storeName )?.getAdBlockerWarningMessage()
	);
	const getHelpLink = useSelect( ( select ) =>
		select( CORE_SITE ).getDocumentationLinkURL(
			`${ moduleSlug }-ad-blocker-detected`
		)
	);

	return (
		<AdBlockerWarningMessage
			className={ className }
			getHelpLink={ getHelpLink }
			warningMessage={ adBlockerWarningMessage }
		/>
	);
}

AdBlockerWarning.propTypes = {
	className: PropTypes.string,
	moduleSlug: PropTypes.string.isRequired,
};
