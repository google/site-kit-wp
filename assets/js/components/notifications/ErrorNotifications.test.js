/**
 * ErrorNotifications component tests.
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import ErrorNotifications from './ErrorNotifications';
import {
	render,
	createTestRegistry,
	provideUserAuthentication,
	provideModules,
	provideSiteInfo,
	provideNotifications,
} from '../../../../tests/js/test-utils';
import {
	CORE_USER,
	FORM_TEMPORARY_PERSIST_PERMISSION_ERROR,
} from '../../googlesitekit/datastore/user/constants';
import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
import { CORE_FORMS } from '../../googlesitekit/datastore/forms/constants';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '../../googlesitekit/constants';
import { READ_SCOPE as TAGMANAGER_READ_SCOPE } from '../../modules/tagmanager/datastore/constants';
import { DEFAULT_NOTIFICATIONS } from '../../googlesitekit/notifications/register-defaults';

describe( 'ErrorNotifications', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();
		provideModules( registry );
		registry.dispatch( CORE_USER ).receiveConnectURL( 'test-url' );
		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );
	} );

	it( 'does not render UnsatisfiedScopesAlert when user is not authenticated', async () => {
		provideUserAuthentication( registry, {
			authenticated: false,
			unsatisfiedScopes: [
				'https://www.googleapis.com/auth/analytics.readonly',
			],
		} );
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideNotifications(
			registry,
			{
				'authentication-error':
					DEFAULT_NOTIFICATIONS[ 'authentication-error' ],
			},
			{ overwrite: true }
		);
		const { container, waitForRegistry } = render( <ErrorNotifications />, {
			registry,
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );
		await waitForRegistry();

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'renders UnsatisfiedScopesAlert when user is authenticated', async () => {
		provideUserAuthentication( registry, {
			grantedScopes: [ TAGMANAGER_READ_SCOPE ],
			unsatisfiedScopes: [
				'https://www.googleapis.com/auth/analytics.readonly',
			],
		} );
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideNotifications(
			registry,
			{
				'authentication-error':
					DEFAULT_NOTIFICATIONS[ 'authentication-error' ],
			},
			{ overwrite: true }
		);

		const { container, waitForRegistry } = render( <ErrorNotifications />, {
			registry,
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );

		await waitForRegistry();

		expect( container ).toHaveTextContent(
			'Site Kit can’t access necessary data'
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'renders `Get help` link', async () => {
		provideUserAuthentication( registry, {
			unsatisfiedScopes: [
				'https://www.googleapis.com/auth/analytics.readonly',
			],
		} );
		provideSiteInfo( registry, {
			proxySupportLinkURL: 'https://test.com',
			setupErrorCode: 'error_code',
			setupErrorMessage: 'An error occurred',
		} );
		const { container, getByRole, waitForRegistry } = render(
			<ErrorNotifications />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);
		await waitForRegistry();

		expect( container ).toHaveTextContent( 'Get help' );
		expect( getByRole( 'link', { name: /get help/i } ) ).toHaveAttribute(
			'href',
			registry.select( CORE_SITE ).getErrorTroubleshootingLinkURL( {
				code: registry.select( CORE_SITE ).getSetupErrorCode(),
			} )
		);
	} );

	it( 'renders the GTE message when the only unsatisfied scope is the tagmanager readonly scope', async () => {
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideUserAuthentication( registry, {
			unsatisfiedScopes: [
				'https://www.googleapis.com/auth/tagmanager.readonly',
			],
		} );
		provideNotifications(
			registry,
			{
				'authentication-error':
					DEFAULT_NOTIFICATIONS[ 'authentication-error' ],
				'authentication-error-gte':
					DEFAULT_NOTIFICATIONS[ 'authentication-error-gte' ],
			},
			{ overwrite: true }
		);

		const { container, waitForRegistry } = render( <ErrorNotifications />, {
			registry,
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );

		await waitForRegistry();

		expect( container ).toHaveTextContent(
			'Site Kit needs additional permissions to detect updates to tags on your site'
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'does not render the GTE message if there are multiple unsatisfied scopes', async () => {
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideUserAuthentication( registry, {
			unsatisfiedScopes: [
				'https://www.googleapis.com/auth/tagmanager.readonly',
				'https://www.googleapis.com/auth/analytics.readonly',
			],
		} );
		provideNotifications(
			registry,
			{
				'authentication-error':
					DEFAULT_NOTIFICATIONS[ 'authentication-error' ],
				'authentication-error-gte':
					DEFAULT_NOTIFICATIONS[ 'authentication-error-gte' ],
			},
			{ overwrite: true }
		);

		const { container, waitForRegistry } = render( <ErrorNotifications />, {
			registry,
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );

		await waitForRegistry();

		expect( container ).toHaveTextContent(
			'Site Kit can’t access necessary data'
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'does render the redo setup CTA if initial Site Kit setup authentication is not granted', async () => {
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideUserAuthentication( registry, {
			unsatisfiedScopes: [
				'https://www.googleapis.com/auth/tagmanager.readonly',
				'https://www.googleapis.com/auth/analytics.readonly',
			],
			authenticated: false,
		} );
		provideNotifications(
			registry,
			{
				'authentication-error':
					DEFAULT_NOTIFICATIONS[ 'authentication-error' ],
				'authentication-error-gte':
					DEFAULT_NOTIFICATIONS[ 'authentication-error-gte' ],
			},
			{ overwrite: true }
		);
		provideSiteInfo( registry, {
			setupErrorRedoURL: '#',
			setupErrorCode: 'access_denied',
			setupErrorMessage:
				'Setup was interrupted because you did not grant the necessary permissions',
		} );

		const { container, waitForRegistry } = render( <ErrorNotifications />, {
			registry,
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );
		await waitForRegistry();

		expect( container ).toHaveTextContent( 'Setup was interrupted' );
		expect( container ).toHaveTextContent( 'Redo the plugin setup' );
	} );

	it( 'does not render the redo setup CTA if it is not due to the interruption of plugin setup and no permission is temporarily persisted', async () => {
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideUserAuthentication( registry );
		provideNotifications(
			registry,
			{
				'authentication-error':
					DEFAULT_NOTIFICATIONS[ 'authentication-error' ],
				'authentication-error-gte':
					DEFAULT_NOTIFICATIONS[ 'authentication-error-gte' ],
			},
			{ overwrite: true }
		);
		provideSiteInfo( registry, {
			setupErrorCode: 'access_denied',
			setupErrorMessage:
				'Setup was interrupted because you did not grant the necessary permissions',
			setupErrorRedoURL: '#',
		} );

		const { container, waitForRegistry } = render( <ErrorNotifications />, {
			registry,
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );
		await waitForRegistry();

		expect( container ).toHaveTextContent( 'Setup was interrupted' );
		expect( container ).not.toHaveTextContent( 'Redo the plugin setup' );
	} );

	it( 'does render the grant permission CTA if additional permissions were not granted and permission is temporarily persisted', async () => {
		provideUserAuthentication( registry );
		provideSiteInfo( registry, {
			isAuthenticated: true,
			setupErrorCode: 'access_denied',
			setupErrorMessage:
				'Setup was interrupted because you did not grant the necessary permissions',
			setupErrorRedoURL: '#',
		} );

		registry
			.dispatch( CORE_FORMS )
			.setValues( FORM_TEMPORARY_PERSIST_PERMISSION_ERROR, {
				status: 403,
				message: 'Generic scope',
				data: {
					scopes: [
						'https://www.googleapis.com/auth/analytics.edit',
					],
				},
			} );

		const { container, waitForRegistry } = render( <ErrorNotifications />, {
			registry,
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );
		await waitForRegistry();

		expect( container ).toHaveTextContent( 'Setup was interrupted' );
		expect( container ).not.toHaveTextContent( 'Grant permission' );
	} );
} );
