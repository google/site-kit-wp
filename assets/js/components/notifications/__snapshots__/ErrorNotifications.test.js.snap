// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorNotifications does not render the GTE message if there are multiple unsatisfied scopes 1`] = `
<div>
  <section
    class="googlesitekit-publisher-win googlesitekit-publisher-win--win-error"
    id="authentication-error"
  >
    <div
      class="mdc-layout-grid"
    >
      <div
        class="mdc-layout-grid__inner"
      >
        <div
          class="googlesitekit-publisher-win__content mdc-layout-grid__cell mdc-layout-grid__cell--span-11-desktop mdc-layout-grid__cell--span-7-tablet mdc-layout-grid__cell--span-3-phone"
        >
          <h3
            class="googlesitekit-heading-2 googlesitekit-publisher-win__title"
          >
            Site Kit can’t access necessary data
          </h3>
          <div
            class="googlesitekit-publisher-win__desc"
          >
            <p>
              <span>
                Site Kit can’t access all relevant data because you haven’t granted all permissions requested during setup. To use Site Kit, you’ll need to redo the setup for: Tag Manager and Analytics – make sure to approve all permissions at the authentication stage.
              </span>
               
            </p>
          </div>
          <a
            class="mdc-button googlesitekit-notification__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
            href="test-url?redirect=http%3A%2F%2Flocalhost%2F"
            role="button"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Redo setup
            </span>
          </a>
        </div>
        <div
          class="mdc-layout-grid__cell mdc-layout-grid__cell--span-1 mdc-layout-grid__cell--order-3-desktop mdc-layout-grid__cell--order-3-tablet mdc-layout-grid__cell--order-3-phone"
        >
          <div
            class="googlesitekit-publisher-win__icons"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`ErrorNotifications renders UnsatisfiedScopesAlert when user is authenticated 1`] = `
<div>
  <section
    class="googlesitekit-publisher-win googlesitekit-publisher-win--win-error"
    id="authentication-error"
  >
    <div
      class="mdc-layout-grid"
    >
      <div
        class="mdc-layout-grid__inner"
      >
        <div
          class="googlesitekit-publisher-win__content mdc-layout-grid__cell mdc-layout-grid__cell--span-11-desktop mdc-layout-grid__cell--span-7-tablet mdc-layout-grid__cell--span-3-phone"
        >
          <h3
            class="googlesitekit-heading-2 googlesitekit-publisher-win__title"
          >
            Site Kit can’t access necessary data
          </h3>
          <div
            class="googlesitekit-publisher-win__desc"
          >
            <p>
              <span>
                Site Kit can’t access the relevant data from Analytics because you haven’t granted all permissions requested during setup. To use Site Kit, you’ll need to redo the setup for Analytics – make sure to approve all permissions at the authentication stage.
              </span>
               
            </p>
          </div>
          <a
            class="mdc-button googlesitekit-notification__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
            href="test-url?redirect=http%3A%2F%2Flocalhost%2F"
            role="button"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Redo setup
            </span>
          </a>
        </div>
        <div
          class="mdc-layout-grid__cell mdc-layout-grid__cell--span-1 mdc-layout-grid__cell--order-3-desktop mdc-layout-grid__cell--order-3-tablet mdc-layout-grid__cell--order-3-phone"
        >
          <div
            class="googlesitekit-publisher-win__icons"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`ErrorNotifications renders the GTE message when the only unsatisfied scope is the tagmanager readonly scope 1`] = `
<div>
  <section
    class="googlesitekit-publisher-win googlesitekit-publisher-win--win-error"
    id="authentication-error-gte"
  >
    <div
      class="mdc-layout-grid"
    >
      <div
        class="mdc-layout-grid__inner"
      >
        <div
          class="googlesitekit-publisher-win__content mdc-layout-grid__cell mdc-layout-grid__cell--span-11-desktop mdc-layout-grid__cell--span-7-tablet mdc-layout-grid__cell--span-3-phone"
        >
          <h3
            class="googlesitekit-heading-2 googlesitekit-publisher-win__title"
          >
            Site Kit needs additional permissions to detect updates to tags on your site
          </h3>
          <div
            class="googlesitekit-publisher-win__desc"
          >
            <p>
              <span>
                To continue using Analytics with Site Kit, you need to grant permission to check for any changes in your Google tag’s target Analytics property. The Google tag feature was recently updated to allow users to change a tag’s connected Analytics property without editing site code. Because of this change, Site Kit now must regularly check if the tag on your site matches the Analytics property destination.
              </span>
               
              <a
                aria-label="Learn more (opens in a new tab)"
                class="googlesitekit-cta-link"
                href="https://support.google.com/tagmanager/answer/11994839?hl=en-US"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Learn more
                </span>
                <span
                  class="googlesitekit-icon-wrapper"
                  style="margin-left: 5px;"
                >
                  <svg />
                </span>
              </a>
            </p>
          </div>
          <a
            class="mdc-button googlesitekit-notification__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
            href="test-url?redirect=http%3A%2F%2Flocalhost%2F&additional_scopes%5B0%5D=gttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Ftagmanager.readonly"
            role="button"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Grant permission
            </span>
          </a>
        </div>
        <div
          class="mdc-layout-grid__cell mdc-layout-grid__cell--span-1 mdc-layout-grid__cell--order-3-desktop mdc-layout-grid__cell--order-3-tablet mdc-layout-grid__cell--order-3-phone"
        >
          <div
            class="googlesitekit-publisher-win__icons"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
